src/ProjectIterator.cpp

```diff
@@ -0,0 +1,41 @@
+//  Copyright (c) 2016, Facebook, Inc.  All rights reserved.
+#include "iterlib/ProjectIterator.h"
+
+namespace iterlib {
+
+bool ProjectIterator::doNext() {
+  return innerIter_->next();
+}
+
+const Item& ProjectIterator::value() const {
```

Comment:
Is it worthwhile to implement a deep clone and fold this functionality into it?
Might make the code less prone to bugs.